pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -684,28 +690,53 @@ public synchronized void clearWarnings() throws SQLException {
     firstWarning = null;
   }
 
-
+  @Override
   public void setReadOnly(boolean readOnly) throws SQLException {
     checkClosed();
+
+    if (readOnly == this.readOnly && cacheXactState) {
```

Comment:
Why did you move that code? It changes semantics, so there should be a reason for the code motion.

Reply:
It not change semantics.

Reply:
It does.
There used to be an exception in case `queryExecutor.getTransactionState() != TransactionState.IDLE)`.
With this change the exception won't be there.

That is a semantic change.

Reply:
The `SET SESSION CHARACTERISTICS AS TRANSACTION` call is omitted if the readOnly is the same, why throw an exception in case `queryExecutor.getTransactionState() != TransactionState.IDLE)`? it do not make any sense.

Sorry but no, this is not a semantic change.